src/TransformableImage.js

```diff
@@ -0,0 +1,205 @@
+    };
+
+    static defaultProps = {
+        enableTransform: true,
+        enableScale: true,
+        enableTranslate: true,
+        imageComponent: undefined,
+        resizeMode: 'contain'
+    };
+
+    constructor (props) {
+        super(props);
+
+        this.onLayout = this.onLayout.bind(this);
+        this.onLoad = this.onLoad.bind(this);
+        this.onLoadStart = this.onLoadStart.bind(this);
+        this.getViewTransformerInstance = this.getViewTransformerInstance.bind(this);
+        this.renderError = this.renderError.bind(this);
+
+        this.state = {
+            viewWidth: 0,
+            viewHeight: 0,
+            imageLoaded: false,
+            imageDimensions: props.source.dimensions,
+            keyAcumulator: 1
+        };
+    }
+
+    componentWillMount () {
+        if (!this.state.imageDimensions) {
+            this.getImageSize(this.props.source);
+        }
+    }
+
+    componentDidMount () {
+        this._mounted = true;
+    }
+
+    componentWillReceiveProps (nextProps) {
+        if (!sameSource(this.props.source, nextProps.source)) {
+            // image source changed, clear last image's imageDimensions info if any
+            this.setState({ imageDimensions: nextProps.source.dimensions, keyAcumulator: this.state.keyAcumulator + 1 });
+            if (!nextProps.source.dimensions) { // if we don't have image dimensions provided in source
+                this.getImageSize(nextProps.source);
+            }
+        }
+    }
+
+    componentWillUnmount () {
+        this._mounted = false;
+    }
+
+    onLoadStart (e) {
+        this.props.onLoadStart && this.props.onLoadStart(e);
+        if (this.state.imageLoaded) {
+            this.setState({ imageLoaded: false });
+        }
+    }
+
+    onLoad (e) {
+        this.props.onLoad && this.props.onLoad(e);
+        if (!this.state.imageLoaded) {
+            this.setState({ imageLoaded: true });
+        }
+    }
+
+    onLayout (e) {
+        let {width, height} = e.nativeEvent.layout;
+        if (this.state.viewWidth !== width || this.state.viewHeight !== height) {
+            this.setState({ viewWidth: width, viewHeight: height });
+        }
+    }
+
+    getImageSize (source) {
+        if (!source) {
+            return;
+        }
+        const { dimensions } = this.props;
+
+        if (dimensions) {
+            this.setState({ imageDimensions: dimensions });
+            return;
+        }
+
+        if (source && source.uri) {
+            Image.getSize(
+                source.uri,
+                (width, height) => {
+                    if (width && height) {
+                        if (this.state.imageDimensions && this.state.imageDimensions.width === width && this.state.imageDimensions.height === height) {
+                            // no need to update state
+                        } else {
+                            this._mounted && this.setState({ imageDimensions: { width, height } });
```

Comment:
setState: { error: false }

Reply:
I usually don't set error to `false` in my code to prevent writing some buggy code when I maintain my libraries.
If I leave it `undefined`, it's still a falsy value, and I'm sure that it will only be evaluated as `true` when I actually wrote `setState({ error: true })`.

It's just a personal preference, though.

Reply:
But after receiving new props if we want to ever re-load again then it will be an error still set,
for first set you're right about undefined value.